doc: adds toplevel docs section (#1207)
tetratelabs/wazero: site/content/docs/_index.md

+++
title = "Docs"
layout = "single"
+++

> ## API

Reviewer: I think overview will be better before anything else. Otherwise, it looks weird to enter with a link like this. I can supply some boilerplate overview if you want.

> Please refer to our [godoc][godoc].

Reviewer: ```suggestion
Being a Go library, which we document wazero's API via [godoc][godoc].
```

+++
title = "Docs"
layout = "single"
+++

## Overview

WebAssembly is a way to safely run code compiled in other languages.
Runtimes execute WebAssembly Modules (Wasm), which are most often binaries with a `.wasm` extension.
Most WebAssembly modules import functions from the host, to perform tasks that are otherwise disallowed by their sandbox.
The most commonly imported functions are called WASI, which allow access to system resources such as the console or files.

wazero is a WebAssembly runtime, written completely in Go. It has no platform dependencies, so can be used in any environment supported by Go.

## API

Being a Go library, which we document wazero's API via [godoc][godoc].

## Terminology

Wazero has consistent terminology used inside the codebase which may be new to you, or different than another WebAssembly runtime.
This section covers the most commonly used vocabulary. Terms rarely used may also be defined inline in individual sections.

* Host - Host is a WebAssembly concept that refers to the process that embeds a WebAssembly runtime. In wazero, the host is a program written in Go.

> * Binary - Binary is a WebAssembly module, compiled from source such as C, Rust or Tinygo. This is also called Wasm or a guest, and usually is a file with a `.wasm` extension. This is the code wazero runs.

Reviewer: ```suggestion
* Binary - Binary is a WebAssembly module, compiled from source such as C, Rust or Tinygo. This is also called Wasm or a guest, and usually is a file with a `.wasm` extension. This is the code wazero runs.
* Sandbox - Sandbox is a term that describes isolation. For example, a WebAssembly module, defined below, is isolated from the host memory and memory of other modules. This means it cannot corrupt the calling process or cause it to crash.
```

> * Module - Module an instance of a Binary, which usually exports functions that can be invoked by the embedder. It can also import functions from the host to perform tasks not defined in the WebAssembly Core Specification, such as I/O.

Reviewer: ```suggestion
* [Module][Module] - Module an instance of a Binary, which usually exports functions that can be invoked by the embedder. It can also import functions from the host to perform tasks not defined in the WebAssembly Core Specification, such as I/O.
```

* Host Module - Host Module is a wazero concept that represents a collection of exported functions that give functionality not provided in the WebAssembly Core Specification, such as I/O. These exported functions are defined as normal Go functions (including closures). For example, WASI is often used to describe a host module named "wasi_snapshot_preview1".
* Exported Function - An Exported Function is a function addressable by name. Guests can import functions from a host module, and export them so that Go applications can call them.
* [Runtime][Runtime] - Runtime is the top-level component in wazero that compiles binaries, configures host functions, and runs guests in sandboxes. How it behaves is determined by its engine: interpreter or compiler.
* Compile - In wazero, compile means prepares a binary, or a host module to be instantiated. This is implemented differently based on whether a runtime is a compiler or an interpreter.
* [Compiled Module][CompiledModule] - a prepared and ready to be instantiated object created vi Compilation phrase. This can be used in instantiation multiple times to create multiple and isolated sandbox from a single Wasm binary.

> * Instantiate - In wazero, instantiate means allocating [a sandbox][SandboxedModule] for a [Compiled Module][CompiledModule] and associating it with a unique name. This includes running any start functions. The result of instantiation is a module whose exported functions can be called.

Reviewer: ```suggestion
* Instantiate - In wazero, instantiate means allocating a [Compiled Module][CompiledModule] and associating it with a unique name, resulting in a [Module][Module]. This includes running any start functions. The result of instantiation is a module whose exported functions can be called.
```

## Architecture

This section covers the library architecture wazero uses to implements the WebAssembly Core specification and WASI.
Features unique to Go or wazero are discussed where architecture affecting.

### Components

At a high level, wazero exposes a [Runtime][Runtime], which can compile the binary into [Compiled Module][CompiledModule],

> and instantiate it into [sandboxed modules][SandboxedModule].

Reviewer: ```suggestion
and instantiate it as sandboxed [Module][Module].
```

the embedding Go programs. During the execution of a function, the objects in the sandbox will be accessed, for example,
a Wasm function can read and write from the memory object in the sandbox. The same goes for globals and tables.

Here's a diagram showing the relationship between these components.

```goat
                                             |           Access during execution
                                             |        +--------+-------+-----------+
                                             |        |        |       |       +---|
                                             |        |        |       |       |   |
                                             |        v        v       v       v   |
                                             |     (Memory, Globals, Table, Functions)
           Wasm Binary                       |                      |              ^
               |                             |                      |              |
+----------+   v   +--------------------+    |    1 : N    +------------------+    |

> | Runtime  | ----> |  Compiled  Module  |----|-----------> | Sandboxed Moudle |    |

Reviewer: ```suggestion
| Runtime  | ----> |  Compiled  Module  |----|-----------> | Module |    |
```

+----------+       +--------------------+    | Instantiate +------------------+    |
                                             |                      |              |
                                             |                      | 1 : N        |
                                             |                      v              |
                                             |             +-------------------+   |
                                             |             | Exported Function |---+
                                             |             +-------------------+
                                             |
                                             |
                           compile time      |      runtime
                                             |
```

> Please note that "Sandboxed Module" is used only for clarity in this document to avoid confusion,

Reviewer: ```suggestion
```

> and it is simply called ["Mudule"][SandboxedModule] at our API level.

Reviewer: ```suggestion
```

#### Compiler

In wazero, a compiler is a runtime configured to compile modules to platform-specific machine code ahead of time (AOT)
during the creation of [CompiledModule][CompiledModule]. This means your WebAssembly functions execute
natively at runtime of the embedding Go program. Compiler is faster than Interpreter, often by order of
magnitude (10x) or more, and therefore enabled by default whenever available.

#### Interpreter

Interpreter is a naive interpreter-based implementation of Wasm virtual machine.
Its implementation doesn't have any platform (GOARCH, GOOS) specific code,
therefore interpreter can be used for any compilation target available for Go (such as riscv64).

## How function call works

How function calls are performed in wazero differ depending on the type of [Engine](#engine).
For those who are curious about how wazero performs function calls under the hood,
please refer to [the dedicated documentation](how_function_call_works)

## Rationales behind wazero

Please refer to [RATIONALE][rationale] for the notable rationales behind wazero's implementations.

> [SandboxedModule]: https://pkg.go.dev/github.com/tetratelabs/wazero@v1.0.0-rc.1/api#Module

Reviewer: ```suggestion
[Module]: https://pkg.go.dev/github.com/tetratelabs/wazero@v1.0.0-rc.1/api#Module
```
